libtenzir/builtins/formats/xsv.cpp

```diff
@@ -248,22 +314,51 @@ auto parse_impl(generator<std::optional<std::string_view>> lines,
       ctrl.warn(
         caf::make_error(ec::parse_error, fmt::format("{} parser skipped line: "
                                                      "parsing line failed",
-                                                     name)));
+                                                     args.name)));
       continue;
     }
     if (values.size() != fields.size()) {
       ctrl.warn(caf::make_error(
         ec::parse_error,
         fmt::format("{} parser skipped line: expected {} fields but got "
                     "{}",
-                    name, fields.size(), values.size())));
+                    args.name, fields.size(), values.size())));
       continue;
     }
     auto row = b.record();
     for (const auto& [field, value] : detail::zip(fields, values)) {
+      // TODO: Instead of the data parser, parse just non-container types, and
+      // respect the list separator and null value options for creating lists.
       auto parsed = data{};
-      if (not(parsers::data - parsers::pattern)(value, parsed)) {
-        parsed = value;
+      const auto value_parser
+        = (parsers::str{args.null_value} >> parsers::eoi).then([](std::string) {
+            return data{};
+          })
```

Comment:
This check treats the quoted null string as a null, which I don't think we want to do?

Reply:
`parsers::str` is not a quoted string, that would be `parsers::qqstr`. For CSV, `"-"` parses as the string `-`, and `-` parses as `null`.

Reply:
Ah, I now see what you mean. There was a double-parsing going on, because we used the parser for the header line to parse the values. I changed this so that we parse the values directly now, and added another test with edge cases in 5fe21af882125c0c569d81eff152f5e262d2532c.

<img width="1667" alt="image" src="https://github.com/tenzir/tenzir/assets/4488655/e01d3cf4-66f3-4d40-b734-3375037cc6d9">

Reply:
Nice, thanks :pray: